File: java/Dson-Core/src/main/java/cn/wjybxx/dson/DsonObjectWriter.java
package cn.wjybxx.dson;

import cn.wjybxx.dson.io.DsonChunk;
import cn.wjybxx.dson.text.INumberStyle;
import cn.wjybxx.dson.text.ObjectStyle;
import cn.wjybxx.dson.text.StringStyle;
import cn.wjybxx.dson.types.ObjectRef;
import cn.wjybxx.dson.types.OffsetTimestamp;

import java.util.Objects;

/**
 * @author wjybxx
 * date - 2023/6/13
 */
public class DsonObjectWriter extends AbstractDsonWriter {

    public DsonObjectWriter(DsonWriterSettings settings, DsonArray<String> outList) {
        super(settings);
        // 顶层输出是一个数组
        Context context = new Context();
        context.init(null, DsonContextType.TOP_LEVEL, null);
        context.container = Objects.requireNonNull(outList);
        setContext(context);
    }

    /** 获取传入的OutList */
    public DsonArray<String> getOutList() {
        Context context = getContext();
        while (context.contextType != DsonContextType.TOP_LEVEL) {
            context = context.getParent();
        }
        return context.container.asArray();
    }

    @Override
    protected Context getContext() {
        return (Context) super.getContext();
    }

    @Override
    protected Context getPooledContext() {
        return (Context) super.getPooledContext();
    }

    @Override
    public void flush() {

    }

    //region 简单值
    @Override
    protected void doWriteInt32(int value, WireType wireType, INumberStyle style) {
        getContext().add(new DsonInt32(value));
    }

    @Override
    protected void doWriteInt64(long value, WireType wireType, INumberStyle style) {
        getContext().add(new DsonInt64(value));
    }

    @Override
    protected void doWriteFloat(float value, INumberStyle style) {
        getContext().add(new DsonFloat(value));
    }

    @Override
    protected void doWriteDouble(double value, INumberStyle style) {
        getContext().add(new DsonDouble(value));
    }

    @Override
    protected void doWriteBool(boolean value) {
        getContext().add(DsonBool.valueOf(value));
    }

    @Override
    protected void doWriteString(String value, StringStyle style) {
        getContext().add(new DsonString(value));
    }

    @Override
    protected void doWriteNull() {
        getContext().add(DsonNull.NULL);
    }

    @Override
    protected void doWriteBinary(DsonBinary binary) {
        getContext().add(binary.copy()); // 需要拷贝
    }

    @Override
    protected void doWriteBinary(int type, DsonChunk chunk) {
        getContext().add(new DsonBinary(type, chunk));
    }

    @Override
    protected void doWriteExtInt32(DsonExtInt32 extInt32, WireType wireType, INumberStyle style) {
        getContext().add(extInt32); // 不可变对象
    }

    @Override
    protected void doWriteExtInt64(DsonExtInt64 extInt64, WireType wireType, INumberStyle style) {
        getContext().add(extInt64);
    }

    @Override
    protected void doWriteExtDouble(DsonExtDouble extDouble, INumberStyle style) {
        getContext().add(extDouble);
    }

    @Override
    protected void doWriteExtString(DsonExtString extString, StringStyle style) {
        getContext().add(extString);
    }

    @Override
    protected void doWriteRef(ObjectRef objectRef) {
        getContext().add(new DsonReference(objectRef));
    }

    @Override
    protected void doWriteTimestamp(OffsetTimestamp timestamp) {
        getContext().add(new DsonTimestamp(timestamp));
    }

    //endregion

    //region 容器
    @Override
    protected void doWriteStartContainer(DsonContextType contextType, DsonType dsonType, ObjectStyle style) {
        Context parent = getContext();
        Context newContext = newContext(parent, contextType, dsonType);
        newContext.container = switch (contextType) {
            case HEADER -> parent.getHeader();
            case ARRAY -> new DsonArray<>();
            case OBJECT -> new DsonObject<>();
            default -> throw new AssertionError();
        };

        setContext(newContext);
        this.recursionDepth++;
    }

    @Override
    protected void doWriteEndContainer() {
        Context context = getContext();
        if (context.contextType != DsonContextType.HEADER) {
            context.getParent().add(context.container);
        }

        this.recursionDepth--;
        setContext(context.parent);
        poolContext(context);
    }
    // endregion

    // region 特殊接口

    @Override
    protected void doWriteValueBytes(DsonType type, byte[] data) {
        throw new UnsupportedOperationException();
    }

    // endregion

    // region context
    private Context newContext(Context parent, DsonContextType contextType, DsonType dsonType) {
        Context context = getPooledContext();
        if (context != null) {
            setPooledContext(null);
        } else {
            context = new Context();
        }
        context.init(parent, contextType, dsonType);
        return context;
    }

    private void poolContext(Context context) {
        context.reset();
        setPooledContext(context);
    }

    protected static class Context extends AbstractDsonWriter.Context {

        DsonValue container;

        public Context() {
        }

        public DsonHeader<String> getHeader() {
            if (container.getDsonType() == DsonType.OBJECT) {
                return container.asObject().getHeader();
            } else {
                return container.asArray().getHeader();
            }
        }

        @SuppressWarnings("unchecked")
        public void add(DsonValue value) {
            if (container.getDsonType() == DsonType.OBJECT) {
                ((DsonObject<String>) container).put(curName, value);
            } else if (container.getDsonType() == DsonType.ARRAY) {
                ((DsonArray<String>) container).add(value);
            } else {
                ((DsonHeader<String>) container).put(curName, value);
            }
        }

        @Override
        public Context getParent() {
            return (Context) parent;
        }
    }
    // endregion
}

File: csharp/Dson/src/DsonObjectWriter.cs
#region LICENSE

//  Copyright 2023 wjybxx(845740757@qq.com)
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to iBn writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#endregion

using System;
using Wjybxx.Dson.IO;
using Wjybxx.Dson.Text;
using Wjybxx.Dson.Types;

#pragma warning disable CS1591
namespace Wjybxx.Dson;

/// <summary>
/// 将对象写入<see cref="DsonArray{TName}"/>
/// </summary>
/// <typeparam name="TName"></typeparam>
public class DsonObjectWriter<TName> : AbstractDsonWriter<TName> where TName : IEquatable<TName>
{
    /// <summary>
    /// 
    /// </summary>
    /// <param name="settings">配置</param>
    /// <param name="outList">接收编码结果</param>
    public DsonObjectWriter(DsonWriterSettings settings, DsonArray<TName> outList)
        : base(settings) {
        // 顶层输出是一个数组
        Context context = new Context();
        context.Init(null, DsonContextType.TopLevel, DsonTypes.Invalid);
        context._container = outList ?? throw new ArgumentNullException(nameof(outList));
        SetContext(context);
    }

    /// <summary>
    /// 获取传入的OutList
    /// </summary>
    public DsonArray<TName> OutList {
        get {
            Context context = GetContext();
            while (context!._contextType != DsonContextType.TopLevel) {
                context = context.Parent;
            }
            return context._container.AsArray<TName>();
        }
    }

    protected override Context GetContext() {
        return (Context)_context;
    }

    protected override Context? GetPooledContext() {
        return (Context)base.GetPooledContext();
    }

    public override void Flush() {
    }

    #region 简单值

    protected override void DoWriteInt32(int value, WireType wireType, INumberStyle style) {
        GetContext().Add(new DsonInt32(value));
    }

    protected override void DoWriteInt64(long value, WireType wireType, INumberStyle style) {
        GetContext().Add(new DsonInt64(value));
    }

    protected override void DoWriteFloat(float value, INumberStyle style) {
        GetContext().Add(new DsonFloat(value));
    }

    protected override void DoWriteDouble(double value, INumberStyle style) {
        GetContext().Add(new DsonDouble(value));
    }

    protected override void DoWriteBool(bool value) {
        GetContext().Add(DsonBool.ValueOf(value));
    }

    protected override void DoWriteString(string value, StringStyle style) {
        GetContext().Add(new DsonString(value));
    }

    protected override void DoWriteNull() {
        GetContext().Add(DsonNull.Null);
    }

    protected override void DoWriteBinary(DsonBinary binary) {
        GetContext().Add(binary.Copy()); // 需要拷贝
    }

    protected override void DoWriteBinary(int type, DsonChunk chunk) {
        GetContext().Add(new DsonBinary(type, chunk));
    }

    protected override void DoWriteExtInt32(DsonExtInt32 extInt32, WireType wireType, INumberStyle style) {
        GetContext().Add(extInt32); // 不可变对象
    }

    protected override void DoWriteExtInt64(DsonExtInt64 extInt64, WireType wireType, INumberStyle style) {
        GetContext().Add(extInt64);
    }

    protected override void DoWriteExtDouble(DsonExtDouble extDouble, INumberStyle style) {
        GetContext().Add(extDouble);
    }

    protected override void DoWriteExtString(DsonExtString extString, StringStyle style) {
        GetContext().Add(extString);
    }

    protected override void DoWriteRef(ObjectRef objectRef) {
        GetContext().Add(new DsonReference(objectRef));
    }

    protected override void DoWriteTimestamp(OffsetTimestamp timestamp) {
        GetContext().Add(new DsonTimestamp(timestamp));
    }

    #endregion

    #region 容器

    protected override void DoWriteStartContainer(DsonContextType contextType, DsonType dsonType, ObjectStyle style) {
        Context parent = GetContext();
        Context newContext = NewContext(parent, contextType, dsonType);
        switch (contextType) {
            case DsonContextType.Header: {
                newContext._container = parent.GetHeader();
                break;
            }
            case DsonContextType.Array: {
                newContext._container = new DsonArray<TName>();
                break;
            }
            case DsonContextType.Object: {
                newContext._container = new DsonObject<TName>();
                break;
            }
            default: throw new InvalidOperationException();
        }

        SetContext(newContext);
        this._recursionDepth++;
    }

    protected override void DoWriteEndContainer() {
        Context context = GetContext();
        if (context._contextType != DsonContextType.Header) {
            context.Parent!.Add(context._container);
        }

        this._recursionDepth--;
        SetContext(context.Parent!);
        PoolContext(context);
    }

    #endregion

    #region 特殊

    protected override void DoWriteValueBytes(DsonType type, byte[] data) {
        throw new InvalidOperationException("Unsupported operation");
    }

    #endregion

    #region context

    private Context NewContext(Context parent, DsonContextType contextType, DsonType dsonType) {
        Context? context = GetPooledContext();
        if (context != null) {
            SetPooledContext(null);
        } else {
            context = new Context();
        }
        context.Init(parent, contextType, dsonType);
        return context;
    }

    private void PoolContext(Context context) {
        context.Reset();
        SetPooledContext(context);
    }

    protected new class Context : AbstractDsonWriter<TName>.Context
    {
#nullable disable
        protected internal DsonValue _container;
#nullable enable

        public Context() {
        }

        public DsonHeader<TName> GetHeader() {
            if (_container.DsonType == DsonType.Object) {
                return _container.AsObject<TName>().Header;
            } else {
                return _container.AsArray<TName>().Header;
            }
        }

        public void Add(DsonValue value) {
            if (_container.DsonType == DsonType.Object) {
                _container.AsObject<TName>().Append(_curName, value);
            } else if (_container.DsonType == DsonType.Array) {
                _container.AsArray<TName>().Add(value);
            } else {
                _container.AsHeader<TName>().Append(_curName, value);
            }
        }

        public new Context Parent => (Context)_parent;
    }

    #endregion
}